### PR TITLE
Implement stratification, stratification-related operators in surface syntax

### PR DESCRIPTION
**Pull request by MingweiSamuel**

### CODE REVIEW

File: hydroflow_lang/src/graph/mod.rs [hunk 1]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+        .collect();
+    let mut node_union: UnionFind<GraphNodeId> = UnionFind::with_capacity(nodes.len());
+    // All edges which belong to a single subgraph. Other & self-edges become handoffs.
+    let mut subgraph_edges: HashSet<(EdgePortRef, EdgePortRef)> = Default::default();
+
+    // Sort edges here (for now, no sort/priority).
+    loop {
+        let mut updated = false;
+        for ((src, src_idx), (dst, dst_idx)) in iter_edges(succs) {
+            // Ignore new self-loops.
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:130]: Comment why we ignore -- is it because they already have been unioned? If so should maybe add an assertion to that effect?

File: hydroflow_lang/src/graph/mod.rs [hunk 2]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
-                            Color::Comp => Color::Push,
-                            Color::Hoff => Color::Pull,
-                            pull_or_push => pull_or_push,
-                        });
-                        updated = true;
-                    }
+            // Set `src` or `dst` color if `None` based on the other (if possible):
+            // Pull -> Pull
+            // Push -> Push
+            // Pull -> [Comp] -> Push
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:151]: What does Comp stand for?

File: hydroflow_lang/src/graph/mod.rs [hunk 3]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+            }
 
-                // If SRC and DST can be in the same subgraph.
-                let can_connect = match (node_color[src], node_color[dst]) {
-                    (Some(Color::Pull), Some(Color::Pull)) => true,
-                    (Some(Color::Pull), Some(Color::Comp)) => true,
-                    (Some(Color::Pull), Some(Color::Push)) => true,
+            // If SRC and DST can be in the same subgraph.
+            let can_connect = match (node_color[src], node_color[dst]) {
+                (Some(Color::Pull), Some(Color::Pull)) => true,
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:176]: With some work I can read that there are 5x5 possibilities and you've handled all of them. I wonder if there's a more readable/checkable way to do that? Possibly encode it as an explicit table we can inspect as text?

Comment (jhellerstein) [reply hydroflow_lang/src/graph/mod.rs:176]: Also factor it out of this routine, which is long.

File: hydroflow_lang/src/graph/mod.rs [hunk 4]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
-            .nodes
-            .keys()
-            .map(|k| (k, Default::default()))
-            .collect();
-
-        // Copy over edges, inserting handoffs between subgraphs (or on subgraph self-edges) when needed.
-        for edge in iter_edges(&flat_graph.succs) {
-            let is_subgraph_edge = subgraph_edges.contains(&edge); // Internal subgraph edges are not handoffs.
-            let ((src, src_idx), (dst, dst_idx)) = edge;
+    // Copy of `self.preds` for the output.
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:204]: I wonder how expensive it is to work with preds and succs, and whether there is a cheaper way to do things without both. Possible add a comment for future benchmarking/perf tuning.

File: hydroflow_lang/src/graph/mod.rs [hunk 5]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+    let mut new_preds: SecondaryMap<GraphNodeId, OutboundEdges> =
+        nodes.keys().map(|k| (k, Default::default())).collect();
+    // Copy of `self.succs` for the output.
+    let mut new_succs: SecondaryMap<GraphNodeId, OutboundEdges> =
+        nodes.keys().map(|k| (k, Default::default())).collect();
+
+    // Copy over edges, inserting handoffs between subgraphs (or on subgraph self-edges) when needed.
+    for edge in iter_edges(succs) {
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:212]: Can we iterate only over the edges between disjoint sets from UnionFind (i.e. the complement of subgraph_edges?) This seems inefficient, but maybe I'm missing something.

File: hydroflow_lang/src/graph/mod.rs [hunk 6]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+    // Determine node's subgraph and subgraph's nodes.
+    // This list of nodes in each subgraph are to be in topological sort order.
+    let (node_subgraph, subgraph_nodes) = {
+        // Ignore handoffs. Esentially each subgraph has it's own topo sort.
+        let topo_sort = graph_algorithms::topo_sort(
+            nodes
+                .iter()
+                .filter(|&(_, node)| !matches!(node, Node::Handoff))
+                .map(|(node_id, _)| node_id),
+            |v| {
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:254]: document why this step is necessary, it's confusing.

File: hydroflow_lang/src/graph/mod.rs [hunk 7]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+    let mut node_color: SecondaryMap<GraphNodeId, Option<Color>> = nodes
+        .keys()
+        .map(|node_id| {
+            let inn_degree = preds[node_id].len();
+            let out_degree = succs[node_id].len();
+            let op_color = node_color(&nodes[node_id], inn_degree, out_degree);
+            (node_id, op_color)
+        })
+        .collect();
+    let mut node_union: UnionFind<GraphNodeId> = UnionFind::with_capacity(nodes.len());
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:122]: name this variable something more meaningful based on its outcome -- basically what is each final disjoint set in your UnionFind? It's a compiled component?

File: hydroflow_lang/src/graph/mod.rs [hunk 8]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+            },
+        );
+
+        // For a `NodeId`, what `SubgraphId` does it belong to.
+        let mut node_subgraph: SecondaryMap<GraphNodeId, GraphSubgraphId> = Default::default();
+        // For a `SubgraphId`, what `NodeId`s belong to it.
+        let mut subgraph_nodes: SlotMap<GraphSubgraphId, Vec<GraphNodeId>> = Default::default();
+        // Populate above.
+        let mut grouped_nodes: SecondaryMap<GraphNodeId, Vec<GraphNodeId>> = Default::default();
+        for node_id in topo_sort {
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:270]: do we have to iterate through `node_id`s? Could we iterate through the disjoint sets from UnionFind and bulk-handle some things? Would that be more efficient?

File: hydroflow_lang/src/graph/mod.rs [hunk 9]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+    barrier_crossers: &[(GraphNodeId, GraphNodeId, IndexInt, InputBarrier)],
+) -> Result<SecondaryMap<GraphSubgraphId, usize>, ()> {
+    // Determine subgraphs's stratum number.
+    // Find SCCs ignoring negative edges, then do TopoSort on the resulting DAG.
+
+    // Generate subgraph graph.
+    let mut subgraph_preds: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = Default::default();
+    let mut subgraph_succs: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = Default::default();
+
+    for (node_id, node) in nodes.iter() {
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:307]: Feels redundant to code above. Would be nice to refactor and also cache if possible.

File: hydroflow_lang/src/graph/mod.rs [hunk 10]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
-                preds: &new_preds,
-                node_union: &mut node_union,
-                marked,
-                subgraph_nodes,
-            };
-            for node_id in flat_graph.nodes.keys() {
-                sg_topo_sort.visit(node_id);
+    let scc = {
+        // https://en.wikipedia.org/wiki/Kosaraju%27s_algorithm
+        fn visit(
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:325]: comment the graph traversal and output order

File: hydroflow_lang/src/graph/mod.rs [hunk 11]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
-                    for &node_id in member_nodes.iter() {
-                        node_subgraph.insert(node_id, subgraph_id);
-                    }
-                    member_nodes
-                });
+        let (mut seen, mut stack) = Default::default();
+        for sg in subgraph_nodes.keys() {
+            visit(&subgraph_succs, sg, &mut seen, &mut stack);
+        }
+        std::mem::drop(seen);
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:343]: This is a little scary. Safe? Necessary?

File: hydroflow_lang/src/graph/mod.rs [hunk 12]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+        for sg in stack.into_iter().rev() {
+            assign(&subgraph_preds, sg, sg, &mut components);
+        }
+
+        components
+    };
+
+    let topo_sort_order = {
+        // Condensed each SCC into a single node for toposort.
+        let condensed_preds: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = subgraph_preds
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:369]: consider materializing on the fly as you find the SCCs?

File: hydroflow_lang/src/graph/mod.rs [hunk 13]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+        })
+    };
+
+    let mut subgraph_stratum: SecondaryMap<GraphSubgraphId, usize> = topo_sort_order
+        .into_iter()
+        .enumerate()
+        .map(|(n, sg_id)| (sg_id, n))
+        .collect();
+
+    let max_stratum = subgraph_stratum.values().cloned().max().unwrap_or(0) + 1;
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:385]: Do we have a philosophy on index-from-1 for strata?

File: hydroflow_lang/src/graph/mod.rs [hunk 14]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+        match input_barrier {
+            InputBarrier::None => unreachable!("No barrier"),
+            InputBarrier::Stratum => {
+                // Any negative edges which go onto the same or previous stratum are bad.
+                // Indicates an unbroken negative cycle.
+                if dst_stratum <= src_stratum {
+                    dst_idx
+                        .span()
+                        .unwrap()
+                        .error("Connection creates a negative cycle which must be broken with a `next_epoch()` operator.")
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:400]: would it be possible for the error message to tell the user where to look in the graph for the semantic bug?

File: hydroflow_lang/src/graph/mod.rs [hunk 15]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
                 }
-                (Node::Handoff, Node::Handoff) => {
-                    Span::call_site().unwrap().error(format!(
-                        "Internal Error: Consecutive handoffs {:?} -> {:?}",
-                        src.data(),
-                        dst.data()
-                    ));
+            }
+            InputBarrier::Epoch => {
+                // If epoch edge goes foreward, need to buffer. (TODO: use a different kind of handoff).
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:406]: What is this TODO about?

File: hydroflow_lang/src/graph/mod.rs [hunk 16]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+                if src_stratum <= dst_stratum {
+                    // Need to inject.
+                    let new_node_id = nodes.insert(Node::Operator(parse_quote! { identity() }));
+                    let new_subgraph_id = subgraph_nodes.insert(vec![new_node_id]);
+                    subgraph_stratum.insert(new_subgraph_id, max_stratum);
+                    node_subgraph.insert(new_node_id, new_subgraph_id);
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:412]: It's kind of a bummer to have to remember all the SecondaryMaps to be handled everywhere. Can we refactor into a single helper?

File: hydroflow_lang/src/graph/mod.rs [hunk 17]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+
+                    let new_hoff_id = nodes.insert(Node::Handoff);
+
+                    {
+                        new_preds.insert(new_node_id, Default::default());
+                        new_succs.insert(new_node_id, Default::default());
+                        new_preds.insert(new_hoff_id, Default::default());
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:419]: why are these lines here twice?

File: hydroflow_lang/src/graph/mod.rs [hunk 18]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+                        new_succs.insert(new_hoff_id, Default::default());
+                    }
+
+                    // Before: A (src) -> H -> B (dst)
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:423]: This code seems unpleasant for a simple insertion to a doubly linked graph. Is there a library for this or can we refactor the graph insertion into subroutines?

File: hydroflow_lang/src/graph/mod.rs [hunk 19]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
-        PartitionedGraph {
+    (subgraph_recv_handoffs, subgraph_send_handoffs)
+}
+
+impl TryFrom<FlatGraph> for PartitionedGraph {
+    type Error = (); // TODO(mingwei).
+
+    fn try_from(mut flat_graph: FlatGraph) -> Result<Self, Self::Error> {
+        // Pairs of node IDs which cross stratums or epochs and therefore cannot be in the same subgraph.
+        let barrier_crossers = iter_edges(&flat_graph.succs)
```
Comment (jhellerstein) [inline hydroflow_lang/src/graph/mod.rs:501]: finding barrier_crossers should probably be a helper function

**Review by jhellerstein — COMMENTED** [review]
First pass through the big algorithmic stuff. Not sure I fully understood it on one pass. Should probably get f2f code review, probably some refactoring of the logic to make it more readable/maintainable.

File: hydroflow_lang/src/graph/mod.rs (hunk 10 above)
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/mod.rs:325]: this should be its own method in `graph_algorithms.rs`

File: hydroflow_lang/src/graph/mod.rs [hunk 20]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+
+        components
+    };
+
+    let topo_sort_order = {
+        // Condensed each SCC into a single node for toposort.
+        let condensed_preds: HashMap<GraphSubgraphId, Vec<GraphSubgraphId>> = subgraph_preds
+            .iter()
+            .map(|(u, preds)| (scc[u], preds.iter().map(|v| scc[v]).collect()))
+            .collect();
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/mod.rs:372]: BUG? overwritten values per key
move into `graph_algorithms.rs`

File: hydroflow_lang/src/graph/mod.rs [hunk 21]
```diff
@@ -85,243 +91,465 @@ pub fn node_color(node: &Node, inn_degree: usize, out_degree: usize) -> Option<C
...
+                            .insert(a_idx, (new_hoff_id, ii0))
+                            .expect("A should be connected to old handoff");
+                        new_preds[new_hoff_id].insert(ii0, (a_id, ii0));
+
+                        //  H' <-> X -> H
+                        new_succs[new_hoff_id].insert(ii0, (new_node_id, ii0));
+                        new_preds[new_node_id].insert(ii0, (new_hoff_id, ii0));
+                        new_succs[new_node_id].insert(ii0, (old_hoff_id, ii0));
+                    }
                 }
```
Comment (MingweiSamuel) [inline hydroflow_lang/src/graph/mod.rs:448]: Put into own fn

File: hydroflow_lang/src/graph/mod.rs (hunk 18 above)
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/mod.rs:423]: will put into own fn

File: hydroflow_lang/src/graph/graph_algorithms.rs [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+) -> Vec<Id>
+where
+    Id: Copy + Hash + Eq,
+    NodeIds: IntoIterator<Item = Id>,
+    PredsFn: FnMut(Id) -> PredsIter,
+    PredsIter: IntoIterator<Item = Id>,
+{
+    let (mut marked, mut order) = Default::default();
+
+    fn visit<Id, PredsFn, PredsIter>(
```
[the remark this answers is not in the document]
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/graph_algorithms.rs:16]: TODO: cycle found return value

File: hydroflow_lang/src/graph/ops/mod.rs [hunk 23]
```diff
@@ -226,37 +302,110 @@ pub const OPERATORS: [OperatorConstraints; 10] = [
...
     //     soft_range_out: RANGE_1,
     //     num_args: 1,
+    //     crosses_stratum: false,
     //     write_prologue_fn: &(|_| quote! {}),
     //     write_fn: &(|_, inputs, outputs, args| {
     //         let ts = quote! { dedup #( #inputs ),* #( #outputs ),* #args };
     //         let lit = Literal::string(&*format!("{}", ts));
     //         quote! { #lit }
     //     }),
     // },
```
[the remark this answers is not in the document]
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/ops/mod.rs:312]: https://github.com/hydro-project/hydroflow/issues/170

File: hydroflow_lang/src/graph/mod.rs (hunk 8 above)
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/mod.rs:270]: keeping as-is

File: hydroflow_lang/src/graph/mod.rs (hunk 11 above)
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/mod.rs:343]: Just early drops the variable, doesnt actually do much. changed to look less scary

File: hydroflow_lang/src/graph/mod.rs (hunk 14 above)
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/mod.rs:400]: Shows it on the negative port edges

File: hydroflow_lang/src/graph/ops/mod.rs [hunk 24]
```diff
@@ -35,21 +46,46 @@ pub struct OperatorConstraints {
...
+pub const OPERATORS: [OperatorConstraints; 18] = [
     OperatorConstraints {
         name: "merge",
         hard_range_inn: &(0..),
         soft_range_inn: &(2..),
         hard_range_out: RANGE_1,
         soft_range_out: RANGE_1,
         num_args: 0,
+        input_barrier_fn: &|_| InputBarrier::None,
         write_prologue_fn: &(|_, _| quote! {}),
```
[the remark this answers is not in the document]
Comment (MingweiSamuel) [reply hydroflow_lang/src/graph/ops/mod.rs:81]: vetoing at least for now
